PYME/LMVis/sessionpaths.py

```diff
@@ -0,0 +1,161 @@
+
+####
+#### BEGIN BLOCK remaining functions for compatibility reading - REMOVE EVENTUALLY
+####
+def fnstring_absolute(fnstring,sessiondir):
+    fnamep = Path(fnstring)
+    if fnamep.is_absolute():
+        fnameabs = fnstring
+    else:
+        fnameabs = str(sessiondir / fnamep) # now make the filenames relative to the session dir path
+    return fnameabs
+
+def get_session_dirP(sessionpath):
+    sessionpathp = Path(sessionpath)
+    abspath = sessionpathp.resolve()
+    sessiondir = abspath.parent # get the directory part of this absolute path
+
+    return sessiondir # return sessiondir as Path object
+    
+
+# note that we do not touch any paths that are already absolute
+# this means we can repeatedly call this on session objects without doing any damage
+# this function only briefly kept for compatibility until all pvs rewritten
+def make_session_paths_absolute_compat(session,sessionpath):
+    chk_fou_modules_registered()
+    sessionabs = session.copy()
+    sessiondir = get_session_dirP(sessionpath)
+    for ds in session['datasources']:
+        fname,query = parse_fnq(session['datasources'][ds])
+        sessionabs['datasources'][ds] = fnq_string(fnstring_absolute(fname,sessiondir),query)
+    # users can request certain recipe module arguments to be also "path-translated" by registering the module/arguments
+    for checkmod in checkmodules:
+        for module in sessionabs['recipe']:
+            if checkmod in module:
+                for field in checkmodules[checkmod]:    
+                    if field in module[checkmod]:
+                        module[checkmod][field] = fnstring_absolute(module[checkmod][field],sessiondir)
+    return sessionabs
+
+####
+#### END BLOCK remaining functions for compatibility reading - REMOVE EVENTUALLY
+####
+
+# split filename?query type string in fname and query parts
+def parse_fnq(dsfnstr):
+    parts = dsfnstr.split('?')
+    parts0 = parts[0]
+    if len(parts) > 1:
+        parts1 = parts[1] # we assume there are no further '?'s in the query string
+    else:
+        parts1 = None
+    return (parts0,parts1)
+
+# join filename and query to form the expected string for session loading
+def fnq_string(fn,query):
+    if query is None:
+        return fn
+    else:
+        return fn + '?' + query
+
+def fnstring_relative(fnstring,sessiondir):
+    fnamep = Path(fnstring)
+    if fnamep.is_absolute():
+        try:
+            fnamerel = str(fnamep.relative_to(sessiondir)) # now make the filenames relative to the session dir path
+        except ValueError: # value error implies it is not possible to make a "clean" relative path, i.e. that the file fnstring is not truely below sessiondir
+            return None # signal via None, we do the further handling in the calling function
+    else:
+        fnamerel = fnstring
+    return fnamerel
+
+def allpaths_relative_to(session,sessiondir):
+    # currently we only check data sources, NOT recipe FileOrURI entries
```

Comment:
We need to check FileOrURI too ....

Reply:
Here it would be good to agree if one of the recipe FileOrURI arguments turning out to be a URI is reason enough to return False.

Am I incorrect in thinking that having a URI in a session to be saved is a bad idea in any case, paths relative or not?

Reply:
I think any paths that are not relative mean that the session is not portable (and it's much safer to be strict here).

I don't see any major issue in having URIs in sessions however (they are portable by default, assuming other settings are correct). In an ideal world, every path in the session would probably be a URI rather than an filesystem path. This assumes that all your computers are set up to use the same cluster(s), but that is the design intent.